### DISCUSSION TITLE
Is there a plan to implement `defaultRefinement={string[]}` in RefinementList for hooks?

### DISCUSSION THREAD

**Discussion by landermedia** [algolia/instantsearch#5312] [Q&A]
In [React InstantSearch RefinementList](https://www.algolia.com/doc/api-reference/widgets/refinement-list/react/), it has optional prop of `defaultRefinement` which automatically selects the item(s) passed by default. However, I can't find this being implemented on [react-instantsearch-hooks-web](https://github.com/algolia/react-instantsearch/blob/master/packages/react-instantsearch-hooks-web/src/ui/RefinementList.tsx).

Question, is there a plan to implement this in the future? 

Also, for now, I'd be interested on how to implement while we're using a custom component that uses `useRefinementList` hook. I've attempted myself with a pretty contrived example below but I don't know what I'm missing here as soon as I update `sortBy` and/or `operator` property, the `defaultRefinement` items no longer gets selected.

```javascript
// sidebar.tsx
const sortByValues: {
  popular?: UseRefinementListProps["sortBy"];
  alphabetical?: UseRefinementListProps["sortBy"];
} = {
  popular: ["isRefined", "count"],
  alphabetical: ["isRefined", "name:asc"],
};

export const SidebarContent = ({ onClose, ...rest }: SidebarProps) => {
  const [refinedTags, setRefinedTags] = React.useState([
    "boy",
    "outdoor",
    "park",
  ]);
  const [operator, setOperator] = React.useState<"and" | "or">("or");
  const [sortByTags, setSortByTags] = React.useState<
    UseRefinementListProps["sortBy"]
  >(sortByValues.popular);
  
  return (
      <>
        <button onClick={() => setSortByTags(sortByValues.popular)}>Popular</button>
        <button onClick={() => setSortByTags(sortByValues.alphabetical)}>A-Z</button>
        <RefinementList
            attribute="_tags"
            operator={operator}
            sortBy={sortByTags}
            defaultRefinement={refinedTags}
        />
      </>
  )
}
```

```javascript
// RefinementList.tsx
export type RefinementListProps = React.ComponentProps<"div"> &
  UseRefinementListProps & {
    defaultRefinement?: string[];
};

function RefinementList(props: RefinementListProps) {
   const {
    items,
    refine,
  } = useRefinementList(props);
  const { defaultRefinement } = props;
  
  React.useEffect(() => {
    if (defaultRefinement?.length > 0) {
      defaultRefinement.forEach((item) => refine(item));
    }
  }, [defaultRefinement]);
  

  console.log(items); // 
  
    // ...more here
}
```

## Actual Result
This works for the first time but as soon as you update `sortBy` prop, the `defaultRefinement` although just the same doesn't seem to do refine once again. 

## Expected Result
Should work even after you update `sortBy` prop.

**Comment by Haroenv** [comment]
We initially don't have that intention, however there's `initialUiState` you can set, however I'm not sure if that solves your use case either, as that likely gets removed once the widget unmounts / mounts again when sortBy is changed. 

In the future we were thinking of making non-destructive prop changes like sortBy not reset the refinement, but for the time being you could do something like getting the ui state before changing the sort by, setting the sort by and immediately setting the previous ui state again.

Some more research in this area needs to be done, but I hope that one of the suggested workarounds I mentioned works for now for you

**Reply by landermedia** [reply]
Thanks @Haroenv. Not sure I'm following but I'll try to see what I can come up with. Thanks for shedding some light on this.

**Comment by ArnaudD** [comment]
Hello @Haroenv,

> In the future we were thinking of making non-destructive prop changes like sortBy not reset the refinement, 

It seems that this is still an issue. Here is a codesandbox demonstrating the problem https://codesandbox.io/s/hooks-example-forked-mdnhne?file=/components/RefinementList.tsx:407-652 Did I do something wrong ?

Is there a GH issue/release I can track regarding this problem ?

> but for the time being you could do something like getting the ui state before changing the sort by, setting the sort by and immediately setting the previous ui state again.

This works indeed 

```ts
const [operator, setOperator] = useState<"and" | "or">("and");
const { items, refine } = useRefinementList({
  attribute: "_tags",
  operator,
});

const { indexUiState, setIndexUiState } = useInstantSearch();

const updateOperator = (newOperator: typeof operator) => {
  const previousTags = indexUiState?.refinementList?._tags;
  setOperator(newOperator);
  setTimeout(() => {
    setIndexUiState((indexUiState) => ({
      ...indexUiState,
      refinementList: {
        ...indexUiState.refinementList,
        _tags: previousTags,
      },
    }));
  }, 0);
};
```

Also, it seems that the operator is not in the indexState and therefore cannot be saved in the URL. Do you have a workaround for this ? 

Thanks a lot

**Reply by Haroenv** [reply]
> Is there a GH issue/release I can track regarding this problem ?

No, you're the first person to request this since the release of React InstantSearch Hooks :) 

The operator is assumed to be static, do you have a UI for that in your page? I guess a custom version of getWidgetUiState / getWidgetSearchParameters could add the info as well, but it doesn't by default indeed

**Reply by ArnaudD** [reply]
Ok, thanks.

You can see it here https://www.finishers.com/courses, at the bottom of the "Thématiques" dropdown (it uses an old version of `react-instantsearch-dom` with `connectRefinementList` HOC for now).

**Reply by Haroenv** [reply]
I see, that's not something we had considered before, I think what would be best is forking the connectRefinementList to make `operator` part of the `getWidgetUiState` / getWidgetSearchParameters` instead of from the parameters. Does that make sense to you?

**Reply by ArnaudD** [reply]
Yes I should be able to do that, thanks !

**Reply by landermedia** [reply]
Following still. Do you mind if you could share how you made it work @ArnaudD? We've put this deep in the backlogs but appreciate if we can move forward with this. 😊

**Reply by ashtonlance** [reply]
@ArnaudD I'm very curious about what solution you were able to find too.

**Reply by ArnaudD** [reply]
Sorry for the late response.

For now our current implementation uses `react-instantsearch-dom`. I haven't tried to implement this feature with `react-instantsearch-hooks` with @Haroenv solution yet.

* Our "Tags filter" component update the url when the user select a different operator. Here is a simplified version :
```js


import { connectRefinementList } from "react-instantsearch-dom";
import { difference } from "lodash";

const TagFilter = connectRefinementList(
  ({ refine, items, currentRefinement = [], operator }) => {
    const router = useRouter();

    const onClick = useCallback(
      (event) => {
        const value = event.currentTarget.getAttribute("data-value");

        if (currentRefinement.includes(value)) {
          refine(difference(currentRefinement, [value]));
        } else {
          refine(currentRefinement.concat(value));
        }
      },
      [refine, currentRefinement]
    );

    const handleOperatorChange = (event) => {
      const value = event.target.value;

      // Update the "tagsOperator" query param in the URL
      router.push(
        {
          pathname: router.pathname,
          query: { ...router.query, tagsOperator: value },
        },
        undefined,
        {
          shallow: true,
        }
      );
    };

    return (
      <div>
        {items.map((choice) => (
          <Item choice={choice} currentRefinement={currentRefinement} onClick={onClick} />
        )}
        <RadioGroup
          value={operator}
          onChange={handleOperatorChange}
        >
          <Radio
            value="and"
            label='AND search'
          />
          <Radio
            value="or"
            label='OR search'
          />
        </RadioGroup>
      </div>
    );
  }
);

TagFilter.defaultProps = {
  limit: 100,
  attribute: "_tags",
  operator: "and",
};

export default TagFilter;
```
* We use [something similar to this](https://github.com/vercel/next.js/blob/bbd79ac9977e03010e09ff90d933f2ec614b031f/examples/with-algolia-react-instantsearch/pages/index.tsx#L52-L58) and [this](https://github.com/vercel/next.js/blob/canary/examples/with-algolia-react-instantsearch/utils/index.ts) to sync the URL with the search state.
```js

const { searchState, onSearchStateChange } = useSearchState(); // This custom hook sync the URL with the search state. 
// The "tagsOperator" state key is synced with the "tagsOperator" query param.

<InstantSearch
  // ...
  searchState={searchState}
  onSearchStateChange={onSearchStateChange}
>

  <TagFilter tagsOperator={searchState.tagsOperator} />
  // ...
</InstantSearch>
```
